packages/datadog-plugin-sharedb/src/index.js

```diff
@@ -0,0 +1,165 @@
+}
+
+function sanitize (input) {
+  const output = {}
+
+  if (!isObject(input) || Buffer.isBuffer(input)) return '?'
+
+  for (const key in input) {
+    if (typeof input[key] === 'function') continue
+
+    output[key] = sanitize(input[key])
+  }
+
+  return output
+}
+
+function isObject (val) {
+  return typeof val === 'object' && val !== null && !(val instanceof Array)
+}
+
+/**
+ * @description Handle a "receive" event. This will be a message to the server, from the client or the same server
+ * process.
+ * @param {Tracer} tracer DataDog Tracer.
+ * @param {sharedb} config sharedb plugin configuration.
+ * @param {string} action ShareDB Message action.
+ * @param {Object} agent ShareDB Agent
+ * @param {Object} triggerContext ShareDB trigger context (internal middleware concept).
+ * @param {Function} callback Callback to continue middleware execution.
+ * @param {Function} triggerFn Function to start middleware execution.
+ * @returns {*}
+ */
+function handleReceive (tracer, config, action, agent, triggerContext, callback, triggerFn) {
+  if (triggerContext.data && triggerContext.data.a) {
+    const scope = tracer.scope()
+    const childOf = scope.active()
+    // Call the trigger function to continue the middleware chain.
+    return triggerFn.call(this, action, agent, triggerContext, function wrappedCallback (err) {
+      // When the middleware calls back into us, start a trace.
+      const actionName = getReadableActionName(triggerContext.data.a)
+      tracer.trace(
+        'sharedb.request',
+        {
+          childOf,
+          tags: {
+            'service.name': config.service || `${tracer._service}-sharedb`,
+            'span.type': 'sharedb.request',
+            'span.kind': 'client',
+            'resource.method': actionName,
+            'resource.name': getReadableResourceName(actionName, triggerContext.data.c, triggerContext.data.q)
+          }
+        },
+        (span, spanDoneCb) => {
+          if (config.hooks && config.hooks.receive) {
+            config.hooks.receive(span, triggerContext)
+          }
+          if (span) {
+            MessagesAwaitingResponse.set(triggerContext.data, {
+              span,
+              spanDoneCb
+            })
+          }
+          callback(err)
+        })
+    })
+  } else {
+    return triggerFn.call(this, action, agent, triggerContext, callback)
+  }
+}
+
+/**
+ * @description Handle a "reply" event. This will be a message to a client for a corresponding "receive" event.
+ * @param {sharedb} config sharedb plugin configuration.
+ * @param {string} action ShareDB Message action.
+ * @param {Object} agent ShareDB Agent
+ * @param {Object} triggerContext ShareDB trigger context (internal middleware concept).
+ * @param {Function} callback Callback to continue middleware execution.
+ * @param {Function} triggerFn Function to start middleware execution.
+ * @returns {*}
+ */
+function handleReply (config, action, agent, triggerContext, callback, triggerFn) {
```

Comment:
Do errors also trigger `reply` events or only `error`?

Reply:
Originally I thought it triggered `send` events, but I couldn't reproduce it. Will investigate more and add tests.